src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -7438,6 +7460,224 @@ private BoundExpression BindIndexedPropertyAccess(SyntaxNode syntax, BoundExpres
             return propertyAccess;
         }
 
+        private bool TryBindIndexOrRangeIndexer(
+            SyntaxNode syntax,
+            BoundExpression receiverOpt,
+            ArrayBuilder<BoundExpression> arguments,
+            out BoundIndexOrRangePatternIndexerAccess patternIndexerAccess)
+        {
+            // Verify a few things up-front, namely that we have a single argument
+            // to this indexer that has an Index or Range type and that there is
+            // a real receiver with a known type
+
+            if (arguments.Count != 1)
+            {
+                patternIndexerAccess = null;
+                return false;
+            }
+
+            var argType = arguments[0].Type;
+            bool argIsIndex = TypeSymbol.Equals(argType,
+                Compilation.GetWellKnownType(WellKnownType.System_Index),
+                TypeCompareKind.ConsiderEverything);
+            bool argIsRange = !argIsIndex || TypeSymbol.Equals(argType,
+                Compilation.GetWellKnownType(WellKnownType.System_Range),
+                TypeCompareKind.ConsiderEverything);
+
+            if ((!argIsIndex && !argIsRange) ||
+                !(receiverOpt?.Type is TypeSymbol receiverType))
+            {
+                patternIndexerAccess = null;
+                return false;
+            }
+
+            // SPEC:
+
+            // An indexer invocation with a single argument of System.Index or System.Range will
+            // succeed if the receiver type conforms to an appropriate pattern, namely
+
+            // 1. The receiver type's original definition has an accessible property getter that returns
+            //    an int and has the name Length or Count
+            // 2. For Index: Has an accessible indexer with a single int parameter
+            //    For Range: Has an accessible Slice method that takes two int parameters
+
+            PropertySymbol lengthOrCountProperty;
+
+            var lookupResult = LookupResult.GetInstance();
+            HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+
+            // look for Length first
+            LookupMembersInType(
+                lookupResult,
+                receiverType,
+                WellKnownMemberNames.LengthPropertyName,
+                arity: 0,
+                basesBeingResolved: null,
+                LookupOptions.Default,
+                originalBinder: this,
+                diagnose: false,
+                useSiteDiagnostics: ref useSiteDiagnostics);
+
+            if (!lookupSatisfiesLengthPattern(out lengthOrCountProperty))
```

Comment:
Nit: Consider a local function that wraps `LookupMembersInType` and `lookupSatfiesLength`. Seems like it would remove some redundancy.